simulator/src/testbench.js

```diff
@@ -73,32 +57,160 @@ export function createTestBenchPrompt() {
+        return;
+    }
+
+}
+
+/**
+ * Updates the TestBench UI on the simulator with the current test attached
+ * If no test is attached then shows the 'No test attached' screen
+ * Called by runTestBench() when test is set, also called by UX/setupPanelListeners()
+ * whenever ux change requires this UI to update(such as clicking on a different circuit or
+ * loading a saved circuit)
+ */
+export function updateTestbenchUI() {
+    // Remove all listeners from buttons
+    $('.tb-dialog-button').off('click');
+    $('.tb-case-button').off('click');
+
+    setupTestbenchUI();
+    if (globalScope.test != undefined) {
+
+        let currentGroup = 0;
+        let currentCase = 0;
+        let result;
+
+        // Initialize the UI
+        setUITableHeaders(globalScope.test);
+
+        // Add listeners to buttons
+        // Previous Case Button
+        $('.tb-case-button#prev-case-btn').on('click', () => {
+            if (currentCase === 0) {
+                if (currentGroup === 0) return;
+                currentGroup--;
+                currentCase = globalScope.test.groups[currentGroup].n - 1;
+            } else currentCase--;
+            setUICurrentCase(globalScope.test, currentGroup, currentCase);
+
+            result = runSingleTest(globalScope.test, currentGroup, currentCase, globalScope);
+            setUIResult(globalScope.test, currentGroup, currentCase, result);
+        });
+
+        // Next Case Button
+        $('.tb-case-button#next-case-btn').on('click', () => {
+            if (currentCase >= globalScope.test.groups[currentGroup].n - 1) {
+                if (currentGroup >= globalScope.test.groups.length - 1) return;
+                currentGroup++;
+                currentCase = 0;
+            } else currentCase++;
+            setUICurrentCase(globalScope.test, currentGroup, currentCase);
+
+            result = runSingleTest(globalScope.test, currentGroup, currentCase, globalScope);
+            setUIResult(globalScope.test, currentGroup, currentCase, result);
+        });
+
+        // Prev Group Button
+        $('.tb-case-button#prev-group-btn').on('click', () => {
+            if (currentGroup === 0) return;
+            currentGroup--;
+            currentCase = 0;
+            setUICurrentCase(globalScope.test, currentGroup, currentCase);
+
+            result = runSingleTest(globalScope.test, currentGroup, currentCase, globalScope);
+            setUIResult(globalScope.test, currentGroup, currentCase, result);
+        });
+
+        // Next Group Button
+        $('.tb-case-button#next-group-btn').on('click', () => {
+            if (currentGroup >= globalScope.test.groups.length - 1) return;
+            currentGroup++;
+            currentCase = 0;
+            setUICurrentCase(globalScope.test, currentGroup, currentCase);
+
+            result = runSingleTest(globalScope.test, currentGroup, currentCase, globalScope);
+            setUIResult(globalScope.test, currentGroup, currentCase, result);
```

Comment:
The above 3 lines seem to be repeated everywhere, can this be improved? Like say we have a function which says
`testbench.setCurrentTestCase(testbench.currentGroup + 1, testbench.currentCase + 1);`
This function itself can validate if that specific action is possible etc and update the UI and run the test accordingly?

I think `currentGroup` and `currentCase` needs to be scoped as a class object to achieve the above.